pandas/io/parsers.py

```diff
@@ -2195,6 +2257,168 @@ def _maybe_parse_dates(self, values, index, try_parse_dates=True):
         return values
 
 
+class BytesIOWrapper:
+    """
+    Allows the pyarrow engine for read_csv() to read from string buffers
+    """
+
+    def __init__(self, string_buffer: StringIO, encoding: str = "utf-8"):
+        self.string_buffer = string_buffer
+        self.encoding = encoding
+
+    def __getattr__(self, attr: str):
+        return getattr(self.string_buffer, attr)
+
+    def read(self, size: int = -1):
+        content = self.string_buffer.read(size)
+        return content.encode(self.encoding)
+
+
+class ArrowParserWrapper(ParserBase):
+    """
+    Wrapper for the pyarrow engine for read_csv()
+    """
+
+    def __init__(self, src, **kwds):
+        self.kwds = kwds
+        self.src = src
+
+        ParserBase.__init__(self, kwds)
+
+        encoding = kwds.get("encoding") if kwds.get("encoding") is not None else "utf-8"
+
+        self.usecols, self.usecols_dtype = _validate_usecols_arg(kwds["usecols"])
+        na_values = kwds["na_values"]
+        if isinstance(na_values, dict):
+            raise ValueError(
+                "The pyarrow engine doesn't support passing a dict for na_values"
+            )
+        self.na_values = list(
+            _clean_na_values(
+                kwds["na_values"], keep_default_na=kwds["keep_default_na"]
+            )[0]
+        )
+        if isinstance(self.src, TextIOBase):
+            self.src = BytesIOWrapper(self.src, encoding=encoding)
+
+    def read(self):
+        VERSIONS["pyarrow"] = "0.15.0"
+        pyarrow = import_optional_dependency(
+            "pyarrow.csv", extra="pyarrow is required to use the pyarrow engine"
+        )
+        kwdscopy = {k: v for k, v in self.kwds.items() if v is not None}
+        # these are kwargs passed to pyarrow
+        parseoptions = {"delimiter", "quote_char", "escape_char", "ignore_empty_lines"}
+        convertoptions = {
+            "include_columns",
+            "null_values",
+            "true_values",
+            "false_values",
+        }
+        parse_options = {k: v for k, v in kwdscopy.items() if k in parseoptions}
+        convert_options = {k: v for k, v in kwdscopy.items() if k in convertoptions}
+        read_options = pyarrow.ReadOptions(autogenerate_column_names=True)
+        headerexists = True if self.header is not None and self.header >= 0 else False
+        skiprows = self.kwds.get("skiprows")
+        if skiprows is not None:
+            read_options = pyarrow.ReadOptions(skip_rows=skiprows)
+        elif headerexists:
+            read_options = pyarrow.ReadOptions(skip_rows=self.header)
+        table = pyarrow.read_csv(
+            self.src,
+            read_options=read_options,
+            parse_options=pyarrow.ParseOptions(**parse_options),
+            convert_options=pyarrow.ConvertOptions(**convert_options),
+        )
+        frame = table.to_pandas()
```

Comment:
Can we somehow allow the user to pass `args/kwargs` to this function call? Especially the `types_mapper` argument would be really useful to support reading into `ExtensionArray` instances.

Reply:
Fully agree, but let's leave that as a follow-up (also for read_parquet this could be useful, although a bit less so given the stored metadata if it is about a roundtrip), and let's first focus on getting the PR merged.

Reply:
Should I make a follow-up issue to track this?

Reply:
Yes, that would be good!

Reply:
https://github.com/pandas-dev/pandas/issues/34823